components/menu/Menu.tsx

```diff
@@ -0,0 +1,172 @@
+import {MenuProps} from './Menu.types';
+
+const Overlay = styled.div(() => {
+    return css`
+        background-color: rgba(0, 0, 0, 0);
+        display: flex;
+        align-items: flex-end;
+        justify-content: center;
+        height: 100%;
+        min-height: 100%;
+        left: 0;
+        right: 0;
+        top: 0;
+        position: fixed;
+        overflow: auto;
+        -webkit-overflow-scrolling: touch;
+        overscroll-behavior: contain;
+        pointer-events: auto;
+        -ms-scroll-chaining: none;
+        transition: 1s;
+        transition-property: background-color;
+        transition-timing-function: ease-in-out;
+
+        &.fade-enter {
+            background-color: rgba(0, 0, 0, 0);
+        }
+
+        &.fade-enter-active {
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &.fade-enter-done {
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &.fade-exit {
+            background-color: rgba(0, 0, 0, 0.5);
+        }
+
+        &.fade-exit-active {
+            background-color: rgba(0, 0, 0, 0);
+        }
+
+        &.fade-exit-done {
+            background-color: rgba(0, 0, 0, 0);
+        }
+    `;
+});
+
+const useIsMobile = (): boolean => {
+    const [isMobile, setIsMobile] = useState<boolean>(false);
+    const mobileMediaQuery = window.matchMedia('(max-width: 699px)');
+
+    const handleChange = (event: MediaQueryListEvent) => {
+        setIsMobile(event.matches);
+    };
+
+    useEffect(() => {
+        mobileMediaQuery.addEventListener('change', handleChange);
+        return () => {
+            mobileMediaQuery.removeEventListener('change', handleChange);
+        };
+    }, [handleChange]);
+
+    useEffect(() => {
+        setIsMobile(mobileMediaQuery.matches);
+    }, []);
+
+    return isMobile;
+};
+
+function usePrevious<T extends unknown>(value: T): T | undefined {
+    // The ref object is a generic container whose current property is mutable
+    // and can hold any value, similar to an instance property on a class
+    const ref = useRef<T>();
+
+    // Store current value in ref
+    useEffect(() => {
+        ref.current = value;
+    }, [value]); // Only re-run if value changes
+    // Return previous value (happens before update in useEffect above)
+
+    return ref.current;
+}
+
+const Menu = (props: MenuProps): JSX.Element => {
+    const {
+        trigger,
+        title,
+        submenuTitle,
+        submenuTrigger,
+        groups,
+        submenuGroups,
+        placement,
+        open,
+        submenuOpen,
+        overlayCloseHandler,
+        closeSubmenu,
+    } = props;
+
+    const menuRef = useRef(null);
+    const submenuRef = useRef(null);
+    const prevOpen = usePrevious(open);
+    const prevSubmenuOpen = usePrevious(submenuOpen);
+    const isMobile = useIsMobile();
+    const isOverlayVisible = isMobile && open;
+
+    useClickOutsideRef([trigger.ref, menuRef, submenuRef], () => {
```

Comment:
I know we talked about this but I wanted to ask a UX person just to make sure we are aligned, @michaelgamble do we want to prevent interaction with the app background when the menu is open?

Reply:
I'd say I'm 3/5 on preventing background interaction. Apart from accidentally interacting with background stuff (which is really bad with click-to-reply turned on), keeping everything looking right when it comes to scrolling is a huge pain. It's nice to see this actually works pretty well in this current version, but when I was playing around with it, it breaks for submenus
![Screen Shot 2022-06-13 at 1 23 18 PM](https://user-images.githubusercontent.com/3277310/173410169-57da9938-c2b4-46b1-9634-7c98ab7ae2db.png)

And even though you can make scrolling work nicely, I'd say I'm even 5/5 about that case that it's not worth the effort. There's a reason why a lot of apps either prevent scrolling with a menu open (Slack, Firefox, Chrome, Notes) or just don't bother making it so that menus move when you scroll c (Discord, Steam).